anvil/src/state.rs

```diff
@@ -381,23 +382,103 @@ impl<BackendData: Backend + 'static> AnvilState<BackendData> {
             seat_name,
             seat,
             start_time: std::time::Instant::now(),
+            last_throttled_frame: None,
             #[cfg(feature = "xwayland")]
             xwayland,
             #[cfg(feature = "xwayland")]
             x11_state: None,
         }
     }
 
-    pub fn send_frames(&self, output: &Output) {
-        self.space.elements().for_each(|window| {
-            if self.space.outputs_for_element(window).contains(output) {
-                window.send_frame(self.start_time.elapsed().as_millis() as u32)
+    pub fn send_frames(&mut self, output: &Output) {
+        let mut opaque_regions: Vec<Rectangle<i32, Logical>> = Vec::new();
+
+        let now = std::time::Instant::now();
+        let send_throttled_frame = self
+            .last_throttled_frame
+            .map(|t| now.duration_since(t) >= std::time::Duration::from_secs(1))
+            .unwrap_or(true);
+
+        if send_throttled_frame {
+            self.last_throttled_frame = Some(now);
+        }
+
+        let layer_map = smithay::desktop::layer_map_for_output(output);
+        let (upper, lower): (Vec<&LayerSurface>, Vec<&LayerSurface>) = layer_map
+            .layers()
+            .rev()
+            .partition(|s| matches!(s.layer(), Layer::Background | Layer::Bottom));
+
+        upper
+            .iter()
+            .filter_map(|layer| layer_map.layer_geometry(layer).map(|geo| (geo.loc, layer)))
+            .for_each(|(location, layer)| {
+                if send_throttled_frame
+                    || !layer.is_obscured_by(
+                        &*opaque_regions
+                            .iter()
+                            .copied()
+                            .map(|mut rect| {
+                                rect.loc -= location;
+                                rect
+                            })
+                            .collect::<Vec<_>>(),
+                    )
+                {
+                    layer.send_frame(self.start_time.elapsed().as_millis() as u32);
+                    opaque_regions.extend(layer.opaque_regions_with_popups().into_iter().map(|mut rect| {
+                        rect.loc += location;
+                        rect
+                    }));
+                }
+            });
+
+        self.space.elements().rev().for_each(|window| {
```

Comment:
What's the reasoning for `rev()`?

Reply:
space.elements returns the elements in back to front order, but the opaque regions check operates front to back